p2p/transport.go

```diff
@@ -0,0 +1,579 @@
+	addr NetAddress,
+	cfg peerConfig,
+) (Peer, error) {
+	c, err := addr.DialTimeout(mt.dialTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(xla): Evaluate if we should apply filters if we explicitly dial.
+	if err := mt.filterConn(c); err != nil {
+		return nil, err
+	}
+
+	sc, ni, err := mt.upgrade(c)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg.outbound = true
+
+	p := mt.wrapPeer(sc, ni, cfg)
+
+	return p, mt.filterPeer(p)
+}
+
+// Close implements transportLifecycle.
+func (mt *MultiplexTransport) Close() error {
+	close(mt.closec)
+
+	return mt.listener.Close()
+}
+
+// Listen implements transportLifecycle.
+func (mt *MultiplexTransport) Listen(addr NetAddress) error {
+	ln, err := net.Listen("tcp", addr.DialString())
+	if err != nil {
+		return err
+	}
+
+	mt.listener = ln
+	mt.nodeAddr = addr
+
+	go mt.acceptPeers()
+
+	return nil
+}
+
+func (mt *MultiplexTransport) acceptPeers() {
+	for {
+		c, err := mt.listener.Accept()
+		if err != nil {
+			// Close has been called so we silently exit.
+			if _, ok := <-mt.closec; !ok {
+				return
+			}
+
+			mt.acceptc <- accept{err: err}
+			return
+		}
+
+		// Connection upgrade and filtering should be asynchronous to avoid
+		// Head-of-line blocking[0].
+		// Reference:  https://github.com/tendermint/tendermint/issues/2047
+		//
+		// [0] https://en.wikipedia.org/wiki/Head-of-line_blocking
+		go func(c net.Conn) {
+			var (
+				ni NodeInfo
+				sc *conn.SecretConnection
+			)
+
+			err := mt.filterConn(c)
+			if err == nil {
+				sc, ni, err = mt.upgrade(c)
+			}
+
+			select {
+			case mt.acceptc <- accept{sc, ni, err}:
+				// Make the upgraded peer available.
+			case <-mt.closec:
+				// Give up if the transport was closed.
+				_ = c.Close()
+				return
+			}
+		}(c)
+	}
+}
+
+func (mt *MultiplexTransport) cleanup(c net.Conn, id ID) error {
+	mt.connsMux.Lock()
+	delete(mt.conns, c.RemoteAddr().String())
+	mt.connsMux.Unlock()
+
+	mt.peersMux.Lock()
+	delete(mt.peers, id)
+	mt.peersMux.Unlock()
+
+	return c.Close()
+}
+
+func (mt *MultiplexTransport) filterConn(c net.Conn) error {
+	// We acquire a single read lock for all filters.
+	mt.connsMux.RLock()
+
+	if _, ok := mt.conns[c.RemoteAddr().String()]; ok {
+		return &ErrRejected{conn: c, isDuplicate: true}
+	}
+
+	for _, f := range mt.connFilters {
+		errc := make(chan error)
+
+		go func(f ConnFilterFunc, c net.Conn, errc chan<- error) {
+			errc <- f(mt.conns, c)
+		}(f, c, errc)
+
+		select {
+		case err := <-errc:
+			return &ErrRejected{conn: c, err: err, isFiltered: true}
+		case <-time.After(mt.filterTimeout):
+			return ErrTransportFilterTimeout
+		}
+	}
+
+	mt.connsMux.RUnlock()
+
+	mt.connsMux.Lock()
+	mt.conns[c.RemoteAddr().String()] = c
+	mt.connsMux.Unlock()
+
+	return nil
+}
+
+func (mt *MultiplexTransport) filterPeer(p Peer) error {
+	// We acquire a single read lock for all filters.
+	mt.peersMux.RLock()
+
+	if _, ok := mt.peers[p.ID()]; ok {
+		return &ErrRejected{id: p.ID(), isDuplicate: true}
```

Comment:
same - peersMux doesn't unlock on these errors

Reply:
Also this mutex madness can be prevented if we move to `PeerSet` and create a similar datatype for the conns, which are threadsafe. Think this is a good direction?